File: nodejs-spec/__tests__/__tests__/naming.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { describe, expect, it } from "vitest";
import {
  topicExchangeName,
  serviceEventQueueName,
  serviceRequestExchangeName,
  serviceResponseExchangeName,
  serviceRequestQueueName,
  serviceResponseQueueName,
  natsStreamName,
  natsSubject,
  translateWildcard,
} from "../src/naming.js";

const fixturesPath = resolve(
  import.meta.dirname,
  "../../../../specification/spec/testdata/naming.json",
);
const fixtures = JSON.parse(readFileSync(fixturesPath, "utf-8"));

describe("naming", () => {
  describe("topicExchangeName", () => {
    for (const tc of fixtures.topicExchangeName) {
      it(`${tc.input.name} -> ${tc.expected}`, () => {
        expect(topicExchangeName(tc.input.name)).toBe(tc.expected);
      });
    }
  });

  describe("serviceEventQueueName", () => {
    for (const tc of fixtures.serviceEventQueueName) {
      it(`${tc.input.exchangeName}, ${tc.input.service} -> ${tc.expected}`, () => {
        expect(
          serviceEventQueueName(tc.input.exchangeName, tc.input.service),
        ).toBe(tc.expected);
      });
    }
  });

  describe("serviceRequestExchangeName", () => {
    for (const tc of fixtures.serviceRequestExchangeName) {
      it(`${tc.input.service} -> ${tc.expected}`, () => {
        expect(serviceRequestExchangeName(tc.input.service)).toBe(
          tc.expected,
        );
      });
    }
  });

  describe("serviceResponseExchangeName", () => {
    for (const tc of fixtures.serviceResponseExchangeName) {
      it(`${tc.input.service} -> ${tc.expected}`, () => {
        expect(serviceResponseExchangeName(tc.input.service)).toBe(
          tc.expected,
        );
      });
    }
  });

  describe("serviceRequestQueueName", () => {
    for (const tc of fixtures.serviceRequestQueueName) {
      it(`${tc.input.service} -> ${tc.expected}`, () => {
        expect(serviceRequestQueueName(tc.input.service)).toBe(tc.expected);
      });
    }
  });

  describe("serviceResponseQueueName", () => {
    for (const tc of fixtures.serviceResponseQueueName) {
      it(`${tc.input.targetService}, ${tc.input.serviceName} -> ${tc.expected}`, () => {
        expect(
          serviceResponseQueueName(
            tc.input.targetService,
            tc.input.serviceName,
          ),
        ).toBe(tc.expected);
      });
    }
  });

  describe("natsStreamName", () => {
    for (const tc of fixtures.natsStreamName) {
      it(`${tc.input.name} -> ${tc.expected}`, () => {
        expect(natsStreamName(tc.input.name)).toBe(tc.expected);
      });
    }
  });

  describe("natsSubject", () => {
    for (const tc of fixtures.natsSubject) {
      it(`${tc.input.stream}, ${tc.input.routingKey} -> ${tc.expected}`, () => {
        expect(natsSubject(tc.input.stream, tc.input.routingKey)).toBe(
          tc.expected,
        );
      });
    }
  });

  describe("translateWildcard", () => {
    for (const tc of fixtures.translateWildcard) {
      it(`${tc.input.routingKey} -> ${tc.expected}`, () => {
        expect(translateWildcard(tc.input.routingKey)).toBe(tc.expected);
      });
    }
  });
});

File: naming_test.go
// MIT License
//
// Copyright (c) 2026 sparetimecoders
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package messaging

import (
	"encoding/json"
	"os"
	"testing"
)

func loadNamingFixtures(t *testing.T) namingFixtures {
	t.Helper()
	data, err := os.ReadFile("testdata/naming.json")
	if err != nil {
		t.Fatalf("failed to read naming fixtures: %v", err)
	}
	var f namingFixtures
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("failed to parse naming fixtures: %v", err)
	}
	return f
}

func TestTopicExchangeName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).TopicExchangeName {
		t.Run(tc.Input.Name, func(t *testing.T) {
			got := TopicExchangeName(tc.Input.Name)
			if got != tc.Expected {
				t.Errorf("TopicExchangeName(%q) = %q, want %q", tc.Input.Name, got, tc.Expected)
			}
		})
	}
}

func TestServiceEventQueueName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).ServiceEventQueueName {
		t.Run(tc.Input.ExchangeName+"_"+tc.Input.Service, func(t *testing.T) {
			got := ServiceEventQueueName(tc.Input.ExchangeName, tc.Input.Service)
			if got != tc.Expected {
				t.Errorf("ServiceEventQueueName(%q, %q) = %q, want %q",
					tc.Input.ExchangeName, tc.Input.Service, got, tc.Expected)
			}
		})
	}
}

func TestServiceRequestExchangeName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).ServiceRequestExchangeName {
		t.Run(tc.Input.Service, func(t *testing.T) {
			got := ServiceRequestExchangeName(tc.Input.Service)
			if got != tc.Expected {
				t.Errorf("ServiceRequestExchangeName(%q) = %q, want %q",
					tc.Input.Service, got, tc.Expected)
			}
		})
	}
}

func TestServiceResponseExchangeName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).ServiceResponseExchangeName {
		t.Run(tc.Input.Service, func(t *testing.T) {
			got := ServiceResponseExchangeName(tc.Input.Service)
			if got != tc.Expected {
				t.Errorf("ServiceResponseExchangeName(%q) = %q, want %q",
					tc.Input.Service, got, tc.Expected)
			}
		})
	}
}

func TestServiceRequestQueueName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).ServiceRequestQueueName {
		t.Run(tc.Input.Service, func(t *testing.T) {
			got := ServiceRequestQueueName(tc.Input.Service)
			if got != tc.Expected {
				t.Errorf("ServiceRequestQueueName(%q) = %q, want %q",
					tc.Input.Service, got, tc.Expected)
			}
		})
	}
}

func TestServiceResponseQueueName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).ServiceResponseQueueName {
		t.Run(tc.Input.TargetService+"_"+tc.Input.ServiceName, func(t *testing.T) {
			got := ServiceResponseQueueName(tc.Input.TargetService, tc.Input.ServiceName)
			if got != tc.Expected {
				t.Errorf("ServiceResponseQueueName(%q, %q) = %q, want %q",
					tc.Input.TargetService, tc.Input.ServiceName, got, tc.Expected)
			}
		})
	}
}

func TestNATSStreamName(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).NATSStreamName {
		t.Run(tc.Input.Name, func(t *testing.T) {
			got := NATSStreamName(tc.Input.Name)
			if got != tc.Expected {
				t.Errorf("NATSStreamName(%q) = %q, want %q", tc.Input.Name, got, tc.Expected)
			}
		})
	}
}

func TestNATSSubject(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).NATSSubject {
		t.Run(tc.Input.Stream+"_"+tc.Input.RoutingKey, func(t *testing.T) {
			got := NATSSubject(tc.Input.Stream, tc.Input.RoutingKey)
			if got != tc.Expected {
				t.Errorf("NATSSubject(%q, %q) = %q, want %q",
					tc.Input.Stream, tc.Input.RoutingKey, got, tc.Expected)
			}
		})
	}
}

func TestTranslateWildcard(t *testing.T) {
	for _, tc := range loadNamingFixtures(t).TranslateWildcard {
		t.Run(tc.Input.RoutingKey, func(t *testing.T) {
			got := TranslateWildcard(tc.Input.RoutingKey)
			if got != tc.Expected {
				t.Errorf("TranslateWildcard(%q) = %q, want %q",
					tc.Input.RoutingKey, got, tc.Expected)
			}
		})
	}
}
